src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -706,11 +747,160 @@ private void EmitLanguageBlockNode(AssociativeNode node)
             throw new NotImplementedException();
         }
 
-        private void EmitRangeExprNode(AssociativeNode node)
+
+
+        private double GetStepValueAsDouble(AssociativeNode stepNode)
         {
-            throw new NotImplementedException();
+            if(stepNode == null)
+            {
+                return 1;
+            }
+            if(stepNode is IntNode stpInt)
+            {
+                return stpInt.Value;
+            }
+            if(stepNode is DoubleNode stpDB)
+            {
+                return stpDB.Value;
+            }
+            return double.NaN;
         }
 
+        private bool CheckIdentType<T>(IdentifierNode ident)
+        {
+            if (variables.TryGetValue(ident.Value, out Tuple<int, Type> output))
+            {
+                if (typeof(T) == output.Item2)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+            private Type EmitRangeExprNode(AssociativeNode node)
+        {
+            //we don't do anything if this is the methodlookup phase
+            //as we need want to access the variable types which are not computed
+            //until the emitIL phase.
+            if (compilePass == CompilePass.MethodLookup)
+            {
+                return null;
+            }
+            const string unselectedToken = "UNSELECTEDTOKEN";
+            const string intRangeMethodName = nameof(Builtin.RangeHelpers.GenerateRangeILInt);
+            const string doubleRangeMethodName = nameof(Builtin.RangeHelpers.GenerateRangeILDouble);
+
+            var range = node as RangeExprNode;
+            var fromNode = range.From;
+            var toNode = range.To;
+            var stepNode = range.Step;
+            var stepOp = range.StepOperator;
+            var hasAmountOperator = range.HasRangeAmountOperator;
+
+            //TODO we may want to do this check again at runtime.
+            if(stepNode is DoubleNode && stepOp == DSASM.RangeStepOperator.Number)
+            {
+                throw new ArgumentException(Resources.kInvalidAmountInRangeExpression);
+            }
+
+
+            var methodName = unselectedToken;
+
+            var isIntStep = stepNode is IntNode stpInt ||
+                (hasAmountOperator && stepOp == DSASM.RangeStepOperator.StepSize && stepNode is DoubleNode stpDB && Math.Truncate(stpDB.Value) == stpDB.Value) ||
```

Comment:
What exactly is the purpose of `hasAmountOperator`?

Reply:
Just so I understand - if we make the condition before the `||` this: `stepNode is intNode && hasAmountOperator && stepOp == DSASM.RangeStepOperator.StepSize`, and the from and to nodes are also integers, then it will always mean generate an integer range, correct?

Reply:
Yes.